system/src/system_connection_manager.cpp

```diff
@@ -0,0 +1,489 @@
+
+    // TODO: Determine a specific interface to bind to, even in the default case.
+    // ie: How should we handle selecting a cloud connection when no interfaces are up/ready?
+    // We should have some historical stats to rely on and then bring that network up? 
+    return bestNetwork;
+}
+
+ConnectionTester::ConnectionTester() {
+    network_interface_t interfaceList[] = { 
+        NETWORK_INTERFACE_ETHERNET,
+#if HAL_PLATFORM_WIFI 
+        NETWORK_INTERFACE_WIFI_STA, 
+#endif
+#if HAL_PLATFORM_CELLULAR
+        NETWORK_INTERFACE_CELLULAR
+#endif
+    };
+    
+    for (const auto& i: interfaceList) {
+        struct ConnectionMetrics interfaceDiagnostics = {};
+        interfaceDiagnostics.interface = i;
+        metrics_.append(interfaceDiagnostics);
+    }
+}
+
+ConnectionTester* ConnectionTester::instance() {
+    static ConnectionTester* tester = new ConnectionTester();
+    return tester;
+}
+
+static int getCloudHostnameAndPort(uint16_t * port, char * hostname, int hostnameLength) {
+    ServerAddress server_addr = {};
+    char tmphost[sizeof(server_addr.domain) + 32] = {};
+    if (hostnameLength < (int)(sizeof(tmphost)+1)) {
+        return SYSTEM_ERROR_TOO_LARGE;
+    }
+
+    HAL_FLASH_Read_ServerAddress(&server_addr);
+    if (server_addr.port == 0 || server_addr.port == 0xFFFF) {
+        server_addr.port = spark_cloud_udp_port_get();
+    }
+
+    system_string_interpolate(server_addr.domain, tmphost, sizeof(tmphost), system_interpolate_cloud_server_hostname);
+    strcpy(hostname, tmphost);
+    *port = server_addr.port;
+
+    //LOG(TRACE, "Cloud hostname#port %s#%d", hostname, *port);
+    return 0;
+};
+
+ConnectionMetrics* ConnectionTester::metricsFromSocketDescriptor(int socketDescriptor) {
+    for (auto& i : metrics_) {
+        if (i.socketDescriptor == socketDescriptor) {
+            return &i;
+        }
+    }
+    return nullptr;
+};
+
+bool ConnectionTester::testPacketsOutstanding() {
+    for (auto& i : metrics_) {
+        if (i.txPacketCount > i.rxPacketCount) {
+            return true;
+        }
+    }
+    return false;
+};
+
+int ConnectionTester::allocateTestPacketBuffers(ConnectionMetrics* metrics) {
+    int maxMessageLength = REACHABILITY_MAX_PAYLOAD_SIZE + sizeof(DTLSPlaintext_t);
+    uint8_t* txBuffer = (uint8_t*)malloc(maxMessageLength);
+    uint8_t* rxBuffer = (uint8_t*)malloc(maxMessageLength);
+
+    if (!(txBuffer && rxBuffer)) {
+        if (txBuffer) {
+            free(txBuffer);
+        }
+        if (rxBuffer) {
+            free(rxBuffer);
+        }
+        LOG(ERROR, "%s failed to allocate connection test buffers of size %d", netifToName(metrics->interface), maxMessageLength);
+        return SYSTEM_ERROR_NO_MEMORY;
+    }
+
+    metrics->txBuffer = txBuffer;
+    metrics->rxBuffer = rxBuffer;
+    return 0;
+};
+
+int ConnectionTester::sendTestPacket(ConnectionMetrics* metrics, int length) {
+    int r = 0;
+    // Only send a new packet if we have received the previous one
```

Comment:
We should still send another packet after some timeout probably.

Reply:
Right now the Internet test / Connection test effectively runs in a busy loop for `REACHABILITY_TEST_DURATION_MS` seconds (currently 2.5s). 
Should the test be run in its own thread to not block the system?

Reply:
I picked an arbitrary 500ms time for sending the next packet. We still block the system, but since this test is only ever run when we lose the cloud connection I think its acceptable for now